pallets/committee/src/lib.rs

```diff
@@ -454,110 +488,67 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
-        ensure!(Self::is_member(&who_id), Error::<T, I>::BadOrigin);
-
-        ensure!(
-            T::EnactProposalMaker::is_pip_id_valid(id),
-            Error::<T, I>::NoSuchProposal
-        );
-
-        // It creates the proposal if it does not exists or vote that proposal.
-        let (aye_call, nay_call) = call_maker();
-        let hash = T::Hashing::hash_of(&aye_call);
-
-        if let Some(voting) = Self::voting(hash) {
-            Self::vote(who_id, hash, voting.index, true)?;
-        } else {
-            Self::propose(who_id, Box::new(aye_call))?;
-        }
-
-        // If voting info has been removed, the vote has finished.
-        // Then we have to clean Pip & the other call.
-        if !<Voting<T, I>>::contains_key(hash) {
-            // Clean proposal
-            let nay_hash = T::Hashing::hash_of(&nay_call);
-
-            <Voting<T, I>>::remove(&nay_hash);
-            <Proposals<T, I>>::mutate(|proposals| proposals.retain(|h| h != &nay_hash));
-        }
-
-        Ok(())
+    fn execute(
+        did: IdentityId,
+        proposal: <T as Trait<I>>::Proposal,
+        hash: T::Hash,
+        ayes: MemberCount,
+        seats: MemberCount,
+    ) {
+        let origin = RawOrigin::Members(ayes, seats).into();
+        let ok = proposal.dispatch(origin).is_ok();
+        Self::deposit_event(RawEvent::Executed(did, hash, ok));
     }
 
     /// Any committee member proposes a dispatchable.
     ///
     /// # Arguments
     /// * `proposal` - A dispatchable call.
-    fn propose(did: IdentityId, proposal: Box<<T as Trait<I>>::Proposal>) -> DispatchResult {
-        // Only committee members can propose
-        ensure!(Self::is_member(&did), Error::<T, I>::BadOrigin);
+    fn propose(
+        origin: <T as frame_system::Trait>::Origin,
+        proposal: <T as Trait<I>>::Proposal,
+    ) -> DispatchResult {
+        // 1. Ensure `origin` is a committee member.
+        let did = Self::ensure_is_member(origin)?;
 
-        // Reject duplicate proposals
+        // 2. Get hash & reject duplicate proposals.
         let proposal_hash = T::Hashing::hash_of(&proposal);
         ensure!(
             !<ProposalOf<T, I>>::contains_key(proposal_hash),
             Error::<T, I>::DuplicateProposal
         );
 
-        // If committee is composed of a single member, execute the proposal
+        // 3. Execute if committee is single member, and otherwise record the vote.
         let seats = Self::members().len() as MemberCount;
         if seats < 2 {
-            let ok = proposal
-                .dispatch(RawOrigin::Members(1, seats).into())
-                .is_ok();
-            Self::deposit_event(RawEvent::Executed(did, proposal_hash, ok));
+            Self::execute(did, proposal, proposal_hash, 1, seats);
         } else {
-            let index = Self::proposal_count();
-            <ProposalCount<I>>::mutate(|i| *i += 1);
+            let index = <ProposalCount<I>>::mutate(|i| mem::replace(i, *i + 1));
```

Comment:
Is there any advantage of using `mem::replace` over a normal `i += 1` when we discard the old value anyway?

Reply:
`index` is used below in `PolymeshVotes { index, ... }` and then emitted as an event. `mem::replace` expresses intent clearer than `{ let old = i; i += 1; old }` imo even though borrowck is fine with either (due to `Copy`).